executor/write.go

```diff
@@ -92,7 +93,11 @@ func updateRecord(ctx context.Context, sctx sessionctx.Context, h kv.Handle, old
 
 	// 3. Compare datum, then handle some flags.
 	for i, col := range t.Cols() {
+		collation := newData[i].Collation()
+		// We should use binary collation to compare datum, otherwise the result will be incorrect.
+		newData[i].SetCollation(charset.CollationUTF8MB4)
```

Comment:
a collation should not be a charset.

Reply:
Do you think it is better to use `utf8mb4_bin` directyly?

Reply:
sorry, I made a mistake

Reply:
why set the collation to `CollationUTF8MB4`?  what if the `newData[i]`'s charset is not `utf8mb4`?

Reply:
We have to use binary collation here to compare the datum, otherwise 'a' and 'A' will be treated as the same. After comparing the datum, we need to set the original collation back to ensure the correctness.

Reply:
right, but if `newData[i]'s` charset is not `utf8mb4`, it seems the charset and collation is mismatched

Reply:
So do you think we should set the collation based on the charset?

Reply:
yes, if you fix the issue that way, but I still feel it is tricky to set a binary collation. Is there another way to solve this problem?

Reply:
I think it may be fine to set `binary`(not `utf8mb4_bin`) directly. Since it is simply a trick here. The mismatch is temporary.

Reply:
`utf8mb4_bin` is not the same as `binary` because of padding.
We should use the `binary` collation.

Reply:
Yes, I have fixed it.